### ISSUE TITLE
New request handler overrides original config path

### ISSUE THREAD

**Issue by bondjimbond** [mjordan/islandora_oai_download_link#5]
When you enable the module, you get two options for request handlers: either "Islandora OAI", or "Islandora OAI with download links". However, the "configure" path is the same for both: admin/islandora/tools/islandora-oai/handler. 

So if you've selected the original, and you try to configure it, you're presented with the config page for the second option. i.e. you have the new fields with download link config options.

Changing the config on this page does appear to have some effect - I changed a site's OAI_ETDMS source datastream from DC to MODS, and the change worked. But the download link is not applied -- I suspect because they've selected the original option. 

Let me know if the above was not entirely clear..

**Comment by mjordan** [comment]
Thanks for catching this. The download-link-specific options should not be showing up if the default hander is selected. I'll take a look.

**Comment by mjordan** [comment]
But, I also have to wonder why this module would be installed if the site was using the default handler.

**Comment by bondjimbond** [comment]
> But, I also have to wonder why this module would be installed if the site was using the default handler.

People are weird. (I wonder why they enabled the module as well, if they aren't going to use it; I suspect they just forgot to select the new handler.)

**Comment by mjordan** [comment]
Yeah, people make software so complicated. Anyway, I'll look into the problem.

**Comment by mjordan** [comment]
OK, now the form is only modified if the current handler is "islandora_oai_download_link". I'd appreciate it if you could test the issue-5 branch to confirm that.

**Comment by bondjimbond** [comment]
Your fix works, but it's still confusing. 

You have two handlers available to select, each of which has a "configure" option.

If you've selected the original handler, clicking Configure beside EITHER of the two takes you to the same page - with no download link config options.

If you've selected the new Download Link handler, clicking Configure beside EITHER of the two takes you to a page WITH download link config options.

If the saw the OAI module works requires a single handler config page, it would make more sense to have a single Configure Handler link, above the field where you select which handler to use. This might require some work on the OAI module itself, though, unless you can override default behaviour.

**Comment by mjordan** [comment]
@bondjimbond I implemented this feature as a separate handler, since I wanted to preserve the default handler, with the assumption that sites that wanted to use the "Islandora OAI with download links" handler would select it as their new default. The UI convention established by this module is that all handlers are listed at the bottom of the main admin form, and the user is expected to select one (via the radio buttons). The fact that the 'configure' link for the handlers that are not selected still appears is a bit misleading, for sure. The original bug you identified in this issue violated the goal of "preserving the default handler" but to honest, I was never, not even once, tempted to click on the 'configure' link for the default handler if the "Islandora OAI with download links" handler was the one that was enabled. Evidence that I am not a typical user, I guess.

We could implement the download links feature so that it is not a separate handler, but instead just an add-on to the default handler. In this case,  sites who want to use the "Islandora OAI with download links" functionality would need to decide to do so not by selecting a different radio button in the handler list but simply by enabling this module. The "configure" admin form would look and work identical to how it does now if the "Islandora OAI with download links" handler is selected.

Which method of deciding to use the "Islandora OAI with download links" handler is better - selecting a radio button (as is the case now) or simply enabling the module? As you point out in [your comment](https://github.com/mjordan/islandora_oai_download_link/issues/5#issuecomment-316780835), at least one site had enabled the module but had forgotten to select the new default handler. No matter which implementation we choose, I have to wonder how confusing either option really would be to the people who would be configuring it, provided it's documented a bit better.

**Comment by bondjimbond** [comment]
Given the way everything else works, the best option may be the "Just enable the module" approach. You could remind the administrator that the module is on when they're configuring the handle by having a note in the new fieldsets that says something like "These additional fields are provided by the Islandora OAI Download Link module. Turn it off to remove them."

**Comment by mjordan** [comment]
OK, let me take a crack at that implementation, and I'll ask you to take a look.

**Comment by mjordan** [comment]
@bondjimbond looking into this, I think I see why I chose to create  separate handler. There's no other way that I can see to intercept the output of the default OAI handler (to add the download links). Specifically, https://github.com/mjordan/islandora_oai_download_link/blob/7.x/includes/handler.inc#L18 takes the output of the default handler and modifies it, and returns the modified output as its own. If we don't intercept the output of the default handler, we'd need to recreate the OAI output from scratch. @jordandukart may have some other ideas.

The only other option I can see is to use custom XSLT transforms to add the links, but I haven't tested that to see if would actually work.

**Comment by jordandukart** [comment]
Without the introduction of an alter or something before writing the XML you'd have to go with the separate handler approach. IIIRC you'll only need to implement the request you are overriding as the rest will fallback to the default handler.

**Comment by bondjimbond** [comment]
Can you set it up so that the new handler becomes selected when you enable the module? That'd at least mitigate the user errors..

**Comment by mjordan** [comment]
@bondjimbond I can when the module is installed, I don't think hooks in .install files fire on enabling (but will check that). What should happen when the module is uninstalled? Should the configured handler always default to the default one?

**Comment by mjordan** [comment]
@jordandukart thanks for the clarification.

**Comment by bondjimbond** [comment]
> What should happen when the module is uninstalled? Should the configured handler always default to the default one?

Probably... I'm not sure what else we'd do.
